### PR TITLE
Fix for issue #15. Add `spacing` option to commentParserToESTree allowing preservation of empty lines in AST parsing.

### PR DESCRIPTION
**Pull request by typhonrt**
As discussed in issue #15 presently `commentParserToESTree` will compact the AST description lines removing empty lines due to coercion of empty `description` strings evaluating to false in this conditional `if (description)` previously on line 365. When the description is the empty string as it is for empty comment lines this block is skipped and the output of `commentParserToESTree` produces compacted comment blocks. This can cause downstream issues when parsing / manipulating full comment blocks with one such example being markdown parsing by subsequent tooling like TypeDoc. 

This PR follows the same style as `comment-parser` introducing a new option `spacing` to `commentParserToESTree` that is by default set to `compact` which mirrors the same functionality as `comment-parser`. Setting this option to `preserve` will maintain empty lines in the parsed AST and subsequent conversion back to a string via `estreeToString`.

Two new tests were added verifying that compacting happens by default and empty lines are preserved when `spacing` is set to `preserve`.

### CODE REVIEW

**Comment by typhonrt** [comment]
OK.. I have thoroughly vetted the new `spacing` option as discussed. In the process of creating the round trip `parseComment` -> `commentParserToESTree` -> `estreeToString` tests I have also adjusted how single / 0th line comments are handled such that they are correctly represented in the round trip processing. 

For multi-line comments the conditional test is now:
```js
    // In `compact` mode skip processing if `description` is an empty string.
    // In `preserve` mode process when `description` is not the empty string
    // or the `delimiter` is not `/**` ensuring empty lines are preserved.
    if ((spacing === 'compact' && description) ||
        (spacing === 'preserve' && (description || delimiter !== '/**'))) {
```

In `preserve` mode the line is processed if description is not the empty string or whenever the delimiter is not the initial leading delimiter. This allows empty lines to be processed and represented in the AST and thus rebuilt with `estreeToString`.

File: src/commentParserToESTree.js
Comment (brettz9) [inline src/commentParserToESTree.js:417]: Is `holder.descriptionLines.length` ever non-zero when the delimiter is `/**`?

Comment (typhonrt) [reply src/commentParserToESTree.js:417]: It is necessary in the single line / 0th end tag section (line 266) where the following test will fail for comment block [4]:

```js
commentBlocks.push(`/**
 * Should parse one line. */`);

// A multi-line comment closing delimiter on last line creates one more line.
// Use `${' '}` to prevent IDEs from removing trailing spaces.
commentResults.push(`/**
 * Should parse one line.${' '}
 */`);
 ```
 
 Without `&& delimiter === '/**'` on line 266 the output is:
 ```js
 /**
   Should parse one line.
  */
  ```
  
In the multi-line tests none of the current tests are affected on line 366 without `&& delimiter === '/**`, but to keep parity w/ line 266 I duplicate the `isDelimiterLine` test. From the investigation I did leading new lines before a comment block are stripped, so the 0th line should always have `/**` as the delimiter.

Comment (typhonrt) [reply src/commentParserToESTree.js:417]: But you are correct that `isDelimiterLine` can just check `const isDelimiterLine = delimiter === '/**'` and no tests fail. So that could be cleaned up dropping the the `holder.descriptionLine.length === 0` check at least with the current tests. If there are more test variations for comment blocks that you can think of it is relatively easy to add them.

File: test/commentParserToESTree.js
Comment (brettz9) [inline test/commentParserToESTree.js]: If you need to get things to work with these changes, I am wondering if there is a bug in `estreeToString` because these particular changes seem odd. If `/**` is being added during stringification already due to the `delimiter` on the `JsdocBlock`, why does this first line have its own `*` delimiter too? What happens if this is instead set to the empty string instead for these (intro) delimiter lines?

Comment (typhonrt) [reply test/commentParserToESTree.js]: You have pointed out the fundamental reason with `JsdocBlock` containing the leading opening delimiter that there is likely an issue in `estreeToString` where a multi-line comment block with the start of the description on the 0th line is not handled correctly and a newline is always inserted after the `/**` from the `JSDocBlock` regardless if there is description content on the 0th line.

I'll take a look at `estreeToString` "later today"... err tomorrow.

What the current changes to `commentParserToESTree` do accomplish w/ the current state of `estreeToString` is create a well formed comment block moving any 0th description line start to the 1st in a multi-line comment. More or less the reason why I asked previously if the intent was to produce well formatted comment blocks on output. 

A choice needs to be made on producing a well formatted comment block or reproduce exactly what was entered. 

Changing the delimter / postDelimiter to the empty string on the first descriptionLine fails this test:
```js
/** It should
 *
 * parse */
```

With my changes / current state produces:
```js
/** 
 * It should
 *
 * parse 
 */
```

With empty strings for "It should" documentLine the result is:
```js
/**
It should
 *
 * parse 
 */ 
 ```

Comment (brettz9) [reply test/commentParserToESTree.js]: Cool. Re: the choice, I think preserving what was entered should be the priority. The AST is meant to be available for allowing manipulation that can bring linting-type changes.

Comment (typhonrt) [reply test/commentParserToESTree.js]: I agree that the intention behind this package is allowing manipulation and not automatic linting-type changes; at least aside from the `compact` / `preserve` spacing.

The changes required to exactly reproduce the input with no ESTree AST manipulation in the round trip processing is certainly expanding the scope of work for this PR. So my initial thought on implementing this is potentially adding two new fields to JsdocBlock and that is the opening / closing delimiter line separator. It will either be the empty string or `\n`. This will be inserted after `/**` before the first document line and before the closing `*/` after the last document line or tag.

This work will require more changes in various tests I'm sure, so is not going to be as "surgical" as the current PR. I'll also add round trip testing for a variety of comment blocks in `compact` spacing mode.

At this point I do want to bring up a change in test tool. I'm sure you are aware of `Vitest`. I also came from the C8 / Mocha angle before switching to Vitest. I temporarily used Vitest when making the changes I did in the current PR. It drastically speeds up testing and ease of evaluating temporary logging in testing through the Vitest UI available that perpetually runs tests quickly when source or tests change. I'm once again going to temporarily install it while I take a look at the work ahead. All of the existing tests work in Vitest. I don't want to rock the boat too much so to speak, but for a project like `jsdoccomment` Vitest is a great fit. I won't commit those changes and move everything back to C8 / Mocha when finished, but it does provide great testing infrastructure for this kind of project. I just want to poll your interest in potentially adopting Vitest though. I think once you see it in action you'd be sold.

Comment (brettz9) [reply test/commentParserToESTree.js]: > The changes required to exactly reproduce the input with no ESTree AST manipulation in the round trip processing is certainly expanding the scope of work for this PR. So my initial thought on implementing this is potentially adding two new fields to JsdocBlock and that is the opening / closing delimiter line separator. It will either be the empty string or `\n`. This will be inserted after `/**` before the first document line and before the closing `*/` after the last document line or tag.

What do you think of having `preserve` mode just include any newlines? I'm fine with it either way though.
 
> This work will require more changes in various tests I'm sure, so is not going to be as "surgical" as the current PR. I'll also add round trip testing for a variety of comment blocks in `compact` spacing mode.

Ok, sure.

> At this point I do want to bring up a change in test tool. I'm sure you are aware of `Vitest`. I also came from the C8 / Mocha angle before switching to Vitest. I temporarily used Vitest when making the changes I did in the current PR. It drastically speeds up testing and ease of evaluating temporary logging in testing through the Vitest UI available that perpetually runs tests quickly when source or tests change. I'm once again going to temporarily install it while I take a look at the work ahead. All of the existing tests work in Vitest. I don't want to rock the boat too much so to speak, but for a project like `jsdoccomment` Vitest is a great fit. I won't commit those changes and move everything back to C8 / Mocha when finished, but it does provide great testing infrastructure for this kind of project. I just want to poll your interest in potentially adopting Vitest though. I think once you see it in action you'd be sold.

Ok, can give it a try, sure.

Comment (typhonrt) [reply test/commentParserToESTree.js]: > What do you think of having `preserve` mode just include any newlines? I'm fine with it either way though.

If you can elaborate on the thought above I'd be glad to receive as much input before getting started. I am going to spend a bit of time to familiarize myself even more with the complete picture of the code base / engineering analysis. I'll likely get started this weekend.

The main goal that we discussed for the upcoming changes is exactly representing the input to output in the `parseComment` -> `commentParserToESTree` -> `estreeToString` round trip. This would be the case for the `preserve` mode. The new line / empty lines will be represented in the `documentLine` array where as the `compact` mode will not have empty `documentLine` entries. 

> > I just want to poll your interest in potentially adopting Vitest though. I think once you see it in action you'd be sold.
> 
> Ok, can give it a try, sure.

Just so there is no shock in test source structure. I normally split out data for data driven tests from the test source itself in a `/test/fixture` directory. To have Mocha or Vitest target the tests I use `.test.js` as the extension for the actual tests. I plan to create `/test/roundTrip.test.js` for the data driven round trip testing w/ the data defined in `/test/fixture/roundTrip.js`. The existing tests will be the same just with a `.test.js` extension.

One more procedural question I have relates to commit count. I try to keep PRs to a minimum, preferably one, commit. This expanded effort is a bit more work, so I'll be doing multiple commits. This can be alleviated with a [squash merge](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github#squashing-your-merge-commits), so it might require a setting change or two.

Comment (brettz9) [reply test/commentParserToESTree.js]: > > What do you think of having `preserve` mode just include any newlines? I'm fine with it either way though.
> 
> If you can elaborate on the thought above I'd be glad to receive as much input before getting started. I am going to spend a bit of time to familiarize myself even more with the complete picture of the code base / engineering analysis. I'll likely get started this weekend.

Sorry, I mean having the `description` just include the initial and final newlines when in `preserve` mode, rather than adding new fields for "opening / closing delimiter line separator".

> The main goal that we discussed for the upcoming changes is exactly representing the input to output in the `parseComment` -> `commentParserToESTree` -> `estreeToString` round trip. This would be the case for the `preserve` mode. The new line / empty lines will be represented in the `documentLine` array where as the `compact` mode will not have empty `documentLine` entries.

I guess you mean `descriptionLines` instead of `documentLine`? I see. I'm fine then with new fields for the opening/closing delimiter, so we don't need to add new entries to the beginning or end of `descriptionLines`.

> > > I just want to poll your interest in potentially adopting Vitest though. I think once you see it in action you'd be sold.
> > 
> > 
> > Ok, can give it a try, sure.
> 
> Just so there is no shock in test source structure. I normally split out data for data driven tests from the test source itself in a `/test/fixture` directory. To have Mocha or Vitest target the tests I use `.test.js` as the extension for the actual tests. I plan to create `/test/roundTrip.test.js` for the data driven round trip testing w/ the data defined in `/test/fixture/roundTrip.js`. The existing tests will be the same just with a `.test.js` extension.

Sure, that's fine.

> One more procedural question I have relates to commit count. I try to keep PRs to a minimum, preferably one, commit. This expanded effort is a bit more work, so I'll be doing multiple commits. This can be alleviated with a [squash merge](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github#squashing-your-merge-commits), so it might require a setting change or two.

Sure, I can do a squash.

Comment (typhonrt) [reply test/commentParserToESTree.js]: > > > What do you think of having `preserve` mode just include any newlines? I'm fine with it either way though.
> > 
> Sorry, I mean having the `description` just include the initial and final newlines when in `preserve` mode, rather than adding new fields for "opening / closing delimiter line separator".

My thinking behind this is informed from the difficulty of working with the `comment-parser` output. `comment-parser` doesn't have the idea of an outer `JsdocBlock` in respect to opening / closing delimiters and for instance will associate the closing delimiter with the last tag processed when present. The main source and duplicated tag source / entries makes it difficult to remove the last tag without also removing the closing delimiter. To not repeat similar difficulties in manipulation co-locating all of the data required to alter the opening / closing delimiters is useful.

The data that needs to be stored also is something that isn't isolated to just the `preserve` mode in respect to opening / closing delimiters.

> Sure, I can do a squash.

Excellent, this will make reviewing / discussing ongoing work much easier as well.

Comment (typhonrt) [reply test/commentParserToESTree.js]: OK.. @brettz9. Everything should be good to go now pending review.

The main core commits are [source changes](https://github.com/es-joy/jsdoccomment/pull/16/commits/9d59a950c9b69e5b74b37fbd96cd7d6affe7e2aa) and [test changes](https://github.com/es-joy/jsdoccomment/pull/16/commits/59b874d296008f4f3cdf767877c22eba43618d24)

I added extensive round trip testing as additional tests covering `parseComment` -> `commentParserToESTree` -> `estreeToString` for all relevant options (`compact` / `preserve` spacing and `preferRawType` / the default pratt type parser for estreeToString). Additionally, I test the `ast.description` and `tag.description` against the `descriptionLines` for each JsdocBlock and tags.

The main addition to the AST generated by `commentParserToESTree` is `delimiterLineBreak` and `preterminalLineBreak` which takes into account all possible variations of starting / closing delimiters. It is either `\n` or an `empty string`. 

I made an update to the README adding these properties.

I did revise `estreeToString` to directly traverse the AST instead of the initial visitor key `args` processing.

This is a non-trivial PR, but is well covered in new tests. Test coverage is now 100% on all files except the utilities in `jsdoccomment.js` which were not utilized in any of the changes.

My general recommendation is bumping to `0.5` as this essentially is a breaking change with the new properties in JsdocBlock and exact support input to output.

I'd be glad to help with any release notes as required.

File: src/index.js
Comment (brettz9) [inline src/index.js]: I'd rather we not export our internal utilities (the `toCamelCase` line). Forces us to make breaking changes if we're just refactoring.

File: test/parseComment.test.js
Comment (brettz9) [inline test/parseComment.test.js]:
```suggestion
    // @ts-expect-error Testing bad arg
```

File: test/parseComment.test.js
Comment (brettz9) [inline test/parseComment.test.js]:
```suggestion
    // @ts-expect-error Testing bad arg
```

File: src/commentParserToESTree.js
Comment (brettz9) [reply src/commentParserToESTree.js:417]: > But you are correct that `isDelimiterLine` can just check `const isDelimiterLine = delimiter === '/**'` and no tests fail. So that could be cleaned up dropping the the `holder.descriptionLine.length === 0` check at least with the current tests.

So will you drop the `holder.descriptionLine.length === 0 ` then?

**Review by brettz9 — CHANGES_REQUESTED** [review]
Just a few tweaks as per my comments. I'll just want to give it a spin against `eslint-plugin-jsdoc` and then I think we may be ready.

**Comment by typhonrt** [comment]
> Just a few tweaks as per my comments. I'll just want to give it a spin against `eslint-plugin-jsdoc` and then I think we may be ready.

Sounds good. I just want to collect all of the changes above in a list:

- Remove internal resources including: `toCamelCase` from exports in the package entry point. What other currently exported functions need to be removed?  In the commit where I changed things to named exports I didn't add anything additional to be exported; just unified the codebase around named exports vs half default / half named, so `toCamelCase` was already exported previously.

- Add comments to any TS directives like `@ts-expect-error`. 

---

Another commit that I'll be getting in is converting `C8` ignore statements in `jsdoccomment.js` to `V8` ignore statements. They are the same and Vitest moved away from C8 in the last ~6 months.  There are also much older `istanbul` coverage ignore statements as well like `// istanbul ignore if`. Both C8 / V8 ignore statements are different than some of the `istanbul` coverage ignore statements and need to be converted into the `/* V8 ignore next 4 */` / line count.

**Comment by brettz9** [comment]
> > Just a few tweaks as per my comments. I'll just want to give it a spin against `eslint-plugin-jsdoc` and then I think we may be ready.
> 
> Sounds good. I just want to collect all of the changes above in a list:
> 
> * Remove internal resources including: `toCamelCase` from exports in the package entry point. What other currently exported functions need to be removed?  In the commit where I changed things to named exports I didn't add anything additional to be exported; just unified the codebase around named exports vs half default / half named, so `toCamelCase` was already exported previously.

Oops, yeah. I think that is the only one that is not JSDoc related, so just that one. The line in the README, "Also currently exports these utilities, though they might be removed in the future:" can probably also be changed to "Also exports these utilities".

> * Add comments to any TS directives like `@ts-expect-error`.
> 
> Another commit that I'll be getting in is converting `C8` ignore statements in `jsdoccomment.js` to `V8` ignore statements. They are the same and Vitest moved away from C8 in the last ~6 months. There are also much older `istanbul` coverage ignore statements as well like `// istanbul ignore if`. Both C8 / V8 ignore statements are different than some of the `istanbul` coverage ignore statements and need to be converted into the `/* V8 ignore next 4 */` / line count.

Ok, sounds good.

**Comment by typhonrt** [comment]
OK.. Everything discussed above should be complete.

The last thing that would be icing on the cake so to speak is producing API docs. As mentioned before the current declarations generated aren't perfect in respect to TypeDoc for some subtle reasons which I can describe. I'm real close to releasing the next version of `esm-d-ts`; actually just waiting on `jsdoccomment` to ship. Perhaps after this PR is merged, but before you release `jsdoccomment` I can work up a new PR that uses `esm-d-ts` to clean up the declarations and produce accurate API docs. You'll be able to evaluate the changes separately from this PR. None of this is required, but I think you'd like the results as it reduces the burden of trying to describe the API in the README.

File: src/commentParserToESTree.js
Comment (brettz9) [reply src/commentParserToESTree.js:417]: There's also still this

File: README.md
Comment (brettz9) [inline README.md]: Could you remove this duplicated line too?

File: README.md
Comment (brettz9) [inline README.md]: And this duplicated line?

Comment (typhonrt) [reply README.md]: I assume you are referring to the duplicated `getTokenizers` / `hasSeeWithLink` lines that previously existed in the README before my changes. I also removed any additional "This might be removed in the future" references.

If there is any more general "housekeeping" tasks beyond the scope of this PR that you'd like to complete it may be easiest to do so yourself after the PR is merged.

Comment (brettz9) [reply README.md]: That's great, thanks.

File: test/commentParserToESTree.js
Comment (brettz9) [reply test/commentParserToESTree.js]: All right. I've reviewed everything and checked against `eslint-plugin-jsdoc`, so I think we're pretty much good to go.

Could you just add to `CHANGES.md`, crediting yourself as `@typhonrt` with the changes? I plan to then do a squash.

Comment (typhonrt) [reply test/commentParserToESTree.js]: Excellent, what version number; `0.43.0`?

Comment (brettz9) [reply test/commentParserToESTree.js]: Yeah `0.43.0` sounds good...

Comment (typhonrt) [reply test/commentParserToESTree.js]: Got it updated. 

Yeah, once you get this merged I'll get another PR together for improved type declarations and API docs. Should have that ready tomorrow for review. One of the nice things you'll see is a more compact way to write `import types` in the JSDoc comments, bundled types, and accurate API docs.

Comment (brettz9) [reply test/commentParserToESTree.js]: > One of the nice things you'll see is a more compact way to write `import types` in the JSDoc comments, bundled types, and accurate API docs.

This is still standard or?

**Review by brettz9 — APPROVED** [review]

**Comment by brettz9** [comment]
I amended the branch after this merge to fix the commit message, so be sure to pull from the latest.

File: test/commentParserToESTree.js
Comment (typhonrt) [reply test/commentParserToESTree.js]: Yeah, it's all standard and you can review those changes separately. 

By default `esm-d-ts` uses the newer `Bundler` option for `moduleResolution` for TSC and you can reference `@typedefs` / type aliases via `import('.').ATYPEDEF` instead of `import('./index.js').ATYPEDEF`. It's small, but nice to reference things without an extension or explicit index file; definitely handy in larger codebases.

The bundled type declaration output is a big win for IDE tooling as all symbols are directly referenced and defined in a single `.d.ts` file rather than `import types` [peppered](https://github.com/es-joy/jsdoccomment/blob/main/dist/parseComment.d.ts#L34) in the raw output of `TSC` currently. Check the various current `.d.ts` files and see `import('./index.js').THING`. I use WebStorm and that won't parse / work correctly with intellisense (see below pict). This works correctly with bundled types.

WebStorm intellisense popover w/ current unbundled types (return type of `JsdocBlockWithInline` does not resolve):
image

**Pull request merged**
